[threads] Added missing initializers, check errors on mutex/cond calls

sshambar opened the pull request:
I struggled with memory corruption problems on OSX for a few hours before finally tracking down some uninitialized thread structures (leading to stack corruption).  In the process of tracking this down, I added checks on all the mutex/condition calls so that the next time this happens, it'll be easier to track down!  This may also catch thread problems in code i haven't tested yet (for example, I don't have a spotify account to run checks on the library, but then they're removing support for it soon, and that's another issue :)

ejurgensen commented:
I’m not so keen on the wrappers, the reasons being:
-	“today wrappers for pthread, tomorrow wrappers for malloc, next week wrappers everywhere”
-	no logging of where the error actually is happening
-	harder to read the code

Had some dialogue on the related issue of wrappers for malloc et al here: https://github.com/ejurgensen/forked-daapd/issues/312

My takeaway is that it would be best to make a macro that checks return value, logs properly and then exits (if the value is bad). Haven’t got around to actually doing it, though.

Could you, for this PR, remove the wrappers and just include the pthread fixes that deal with the memory corruption that you ran into?

sshambar commented:
I'd hate to lose the checking... worked really well on finding the corruption, and will catch incorrect nested/x-thread locking issues in other code too.

I can create macros that pass __FILE__/__LINE__ and add logging before an abort (these calls "should" never fail), but I'd suggest keeping the underlying functions just to minimize the code bloat (but could move it all into a macro if you prefer...)

Re: reproducible builds from the discussion you linked: newer clang and gcc are getting support for SOURCE_PREFIX_MAP which normalize the __FILE__ values... but if we're going that route, we'd need to modify the build-date to use SOURCE_DATE_EPOCH too...

sshambar commented:
Something like this:
```c
int fork_cond_timedwait(pthread_cond_t *cond, pthread_mutex_t *mutex,
  const struct timespec *ts, const char *file, int line);
#define FORK_COND_TIMEDWAIT(cond, mutex, ts) \
  fork_cond_timedwait((cond), (mutex), (ts), __FILE__, __LINE__)
```
... with detailed logging in the implementation.  This way it's clear it's a macro, easy to set breakpoints, and avoids code duplication.

ejurgensen commented:
Sorry for the delayed response on this. I still find it hard to justify why specifically pthread_xxx should have wrappers, so I don't want those. I prefer something (probably a generic macro, possibly calling a function) that tests return values of all "unlikely to fail"-functions.

sshambar commented:
Something along these lines?:
```c
#define CHECK_ERR(f) \
  do { int err; err = f; if (err != 0) { \
      printf(#f " failed at %s:%d, err %d (%s)\n", __FILE__, __LINE__, err, \
             strerror(err)); \
      abort(); \
    } } while(0)

CHECK_ERR(pthread_mutex_lock(&t));
````
which would produce error reports like:
pthread_mutex_lock(&t) failed at main.c:33, err 22 (Invalid argument)

I think a wrapper for pthread_mutex_init might still be useful since it requires allocating/freeing an attribute to setup error checking... which is quite verbose :)

ejurgensen commented:
Yes, very neat. Would probably need to have two variants, one like this, and one aborting on null returns (for malloc et al). The error message should go to the log file.

Not quite sure how much compilers optimize, but if you think this will bloat is it then better to let the macro expand to a function in misc.h/.c?

ejurgensen commented:
> I think a wrapper for pthread_mutex_init might still be useful since it requires allocating/freeing an attribute to setup error checking... which is quite verbose :)

Yes that's true. The pthread people should really have made a convenience function if multiple functions are required for a standard init, but in lack of that I'm ok with a wrapper.

sshambar commented:
The bulk of the size is in the new string literals for logging with the macro... I could reduce that by putting the common code in misc:
```c
#include <pthread.h>
int fork_log_fatal(const char *func_name, const char *file, int line, int err) {
  DPRINTF(L_MISC, E_FATAL, "%s failed at %s:%d, err %d (%s)\n", func_name,
          file, line, err, strerror(err));
  abort();
}

#define CHECK_ERR(f) \
  do { int err; err = f; if (err != 0) { \
  fork_log_fatal(#f, __FILE__, __LINE__, err); \
    } } while(0)
```

ejurgensen commented:
Sounds good to me. Just call it log_fatal or something - the prefix fork_ is not used in forked-daapd.

To reduce size, I suppose you could omit the \_\_FILE\_\_ and instead use the log domain. Function, log domain and line number is plenty to locate the culprit. However, that would involve making a few changes to the way the log domain is defined, so maybe it is a change better saved for later.

sshambar commented:
```c
void log_fatal(int domain, const char *func, int line, int err) {
  DPRINTF(E_FATAL, domain, "%s failed at line %d, err %d (%s)\n", func,
          line, err, strerror(err));
  abort();
}

#define CHECK_ERR(d, f) \
  do { int err; err = f; if (err != 0) { \
      log_fatal(d, #f, __LINE__, err); \
    } } while(0)                                 
CHECK_ERR(L_MDNS, pthread_mutex_lock(&t)); 
```
Produces:
[FATAL]     mdns: pthread_mutex_lock(&t) failed at line 34, err 22 (Invalid argument)

Look good?

ejurgensen commented:
Beautiful!

sshambar commented:
I added 4 macros, one to check for NULL, one to check returned error != 0, one to check err < 0 (and use errno in logging), and one to whitelist at least one "expected" error (used for cond_timedwait).  I updated all the mutex/cond calls updated to use macros.
In addition, I updated the logging to use special macros to avoid nested logging issues, and handle early logging (before logging initialized).

ejurgensen commented on src/misc.c:
```diff
@@ -33,6 +33,7 @@
 #include <stdint.h>
 #include <limits.h>
 #include <sys/param.h>
+#include <assert.h>
```
Is this required now?

sshambar replied:
Nope, I'll remove it

ejurgensen commented on src/spotify.c:
```diff
@@ -2042,7 +2044,7 @@ void
 {
   const char *code;
-  const char *err;
+  const char *err = "";
```
Is this required? If there are code paths where spotifywebapi_token_get() returns < 0 without setting err it should be fixed. For safety, spotifywebapi_token_get() should probably also set err to "" to start with.

sshambar replied:
This was to silence a compiler warning (which actually went away when the called function moved to another file with the merge with master...). It's still a good idea for the called function to init err to avoid random errors if the token calls are changed later, I'll move it in there...

sshambar commented:
OK, changes applied.

ejurgensen commented:
Sorry, just one final thing! There are some deletions of include <pthread.h>that shouldn't be there, now that it isn't all wrapped. Would you mind putting them back?

sshambar commented:
I could, although they'd be duplicates, as misc.h already includes pthread.h for the init_mutex api... let me know if you still want me to add them.

ejurgensen commented:
True! Nevermind then. Merged, thanks a lot.

ejurgensen commented:
Seems there is another issue - I get the following build warnings:

```
logger.c: In function ‘vlogger’:
logger.c:148:3: warning: implicit declaration of function ‘abort’ [-Wimplicit-function-declaration]
   LOGGER_CHECK_ERR(pthread_mutex_lock(&logger_lck));
   ^
logger.c:55:7: warning: incompatible implicit declaration of built-in function ‘abort’ [enabled by default]
       abort(); \
       ^
logger.c:148:3: note: in expansion of macro ‘LOGGER_CHECK_ERR’
   LOGGER_CHECK_ERR(pthread_mutex_lock(&logger_lck));
   ^
logger.c:55:7: warning: incompatible implicit declaration of built-in function ‘abort’ [enabled by default]
       abort(); \
       ^
logger.c:152:7: note: in expansion of macro ‘LOGGER_CHECK_ERR’
       LOGGER_CHECK_ERR(pthread_mutex_unlock(&logger_lck));
       ^
logger.c:55:7: warning: incompatible implicit declaration of built-in function ‘abort’ [enabled by default]
       abort(); \
       ^
logger.c:158:3: note: in expansion of macro ‘LOGGER_CHECK_ERR’
   LOGGER_CHECK_ERR(pthread_mutex_unlock(&logger_lck));
```

.... and so on. Can you look into it?
